### ISSUE TITLE
Need example of how to upcast so nested view element attributes get added to model

### ISSUE THREAD

**Issue by cordesmj** [ckeditor/ckeditor5#17209]
### Origin URL

https://ckeditor.com/docs/ckeditor5/latest/framework/deep-dive/conversion/upcast.html

### Project version

43.1.0

### Is there something missing in the guide? What is it?

I am trying to create an upcast conversion, but the provided example doesn't seem to have everything I need.

I want to create a model, `myimage`, to accommodate both images as well as images wrapped with an anchor tag. I want to to copy attributes 'alt' and 'src' from the `<img>` to `myimage`, and if it is wrapped with an anchor, I also want to copy 'href' and 'target' from the `<a>` to `myimage`.

I am trying to accomplish each case with a separate handler, the first for a plain image, and the second for a link wrapping an image, like the following:
```
      schema.register('myimage', {
        inheritAllFrom: '$inlineObject',
        allowAttributes: [ 'alt', 'class', 'src', 'style', 'href', 'target' ],
        allowChildren: false,
        isContent: true,
        isSelectable: true
      });

      editor.conversion.attributeToAttribute({ view: 'alt', model: 'alt' });
      editor.conversion.attributeToAttribute({ view: 'class', model: 'class' });
      editor.conversion.attributeToAttribute({ view: 'src', model: 'src' });
      editor.conversion.attributeToAttribute({ view: 'style', model: 'style' });

      editor.conversion.attributeToAttribute({ view: 'href', model: 'href' });
      editor.conversion.attributeToAttribute({ view: 'target', model: 'target' });

      editor.conversion.for( 'upcast' ).add( dispatcher => {
        // Define elements consumables
        const imageWrapper = { name: true, attributes: ['src'] };
        const linkWrapper = { name: true, attributes: ['href'] };

        // Look for every view <img> element.
        dispatcher.on('element:img', (evt, data, conversionApi) => {
          const {
            consumable,
            writer,
            safeInsert,
            updateConversionResult
          } = conversionApi;
          const image = data.viewItem;

          // Images with links will be inserted by the next handler
          if (image.parent.name === 'a') return;

          // Tests if the view element can be consumed
          if (!consumable.test(image, imageWrapper)) return;

          // Create model element
          const modelElement = writer.createElement( 'myimage' );

          // Insert element on a current cursor location
          if (!safeInsert(modelElement, data.modelCursor)) return;

          // Consume the element
          consumable.consume(image, imageWrapper);
          consumable.add(image, { attributes: ['alt', 'src'] });
          updateConversionResult(modelElement, data);
        });

        // Look for every view <a> element.
        dispatcher.on('element:a', (evt, data, conversionApi) => {
          const {
            consumable,
            writer,
            safeInsert,
            updateConversionResult
          } = conversionApi;
          const link = data.viewItem;
          const linkImage = link.getChild(0);

          // Links that do not wrap an image will be ignored by this plugin
          if (!linkImage?.is('element', 'img')) return;

          // Tests if the view elements can be consumed
          if (!consumable.test(link, linkWrapper)) return;
          if (!consumable.test(linkImage, imageWrapper)) return;

          // Create model element
          const modelElement = writer.createElement('myimage');

          // Insert element on a current cursor location
          if (!safeInsert(modelElement, data.modelCursor)) return;

          // Consume the element
          consumable.consume(link, linkWrapper);
          consumable.consume(linkImage, imageWrapper);
          consumable.add(link, { name: true, attributes: ['href', 'target', ] });
          consumable.add(linkImage, { name: true, attributes: ['alt', 'src'] });

          updateConversionResult(modelElement, data);
        });
      });
    }
```

The `element:img` handler seems to work - I see the model element in the ckeditor dev tool, with the expected attributes. The `element:a` handler, though, only has the attributes from the `<a>` tag, it doesn't include the attributes from the `<img>` on the `myimage` model element.

This case seems to be exactly what the event-based conversion API should be used for, but I think the example is missing something. Specifically, how can I add attributes from nested view elements to the model?

### User agent

Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36

**Comment by cordesmj** [comment]
I was able to get what I want with the following, but man, it seems like a long way to go for something that seems pretty simple. Is there some better way to do this?
```
      schema.register('myimage', {
        inheritAllFrom: '$inlineObject',
        allowAttributes: [ 'alt', 'class', 'src', 'style', 'href', 'target' ],
        allowChildren: false,
        isContent: true,
        isSelectable: true
      });

      editor.conversion.attributeToAttribute({ view: 'href', model: 'href' });
      editor.conversion.attributeToAttribute({ view: 'target', model: 'target' });

      editor.conversion.for('upcast').attributeToAttribute({ view: 'alt', model: 'alt' });
      editor.conversion.for('upcast').attributeToAttribute({ view: 'class', model: 'class' });
      editor.conversion.for('upcast').attributeToAttribute({ view: 'src', model: 'src' });
      editor.conversion.for('upcast').attributeToAttribute({ view: 'style', model: 'style' });

      editor.conversion.for('dataDowncast').attributeToAttribute({
        view: { name: 'img', key: 'alt' },
        model: 'alt'
      });
      editor.conversion.for('dataDowncast').attributeToAttribute({
        view: { name: 'img', key: 'class' },
        model: 'class'
      });
      editor.conversion.for('dataDowncast').attributeToAttribute({
        view: { name: 'img', key: 'src' },
        model: 'src'
      });
      editor.conversion.for('dataDowncast').attributeToAttribute({
        view: { name: 'img', key: 'style' },
        model: 'style'
      });

      editor.conversion.for( 'upcast' ).add( dispatcher => {
        // Define elements consumables
        const wrapper = { name: true };

        // Look for every view <img> element.
        dispatcher.on('element:img', (evt, data, conversionApi) => {
          const {
            consumable,
            writer,
            safeInsert,
            updateConversionResult
          } = conversionApi;
          const image = data.viewItem;

          // Linked images will be inserted by the next handler
          if (image.parent.name === 'a') return;

          // Tests if the view element can be consumed
          if (!consumable.test(image, wrapper)) return;

          // Create model element
          const modelElement = writer.createElement( 'myimage' );

          // Insert element on a current cursor location
          if (!safeInsert(modelElement, data.modelCursor)) return;

          // Consume the element
          consumable.consume(image, wrapper);
          consumable.add(image, { attributes: ['alt', 'src'] });
          updateConversionResult(modelElement, data);
        });

        // Look for every view <a> element.
        dispatcher.on('element:a', (evt, data, conversionApi) => {
          const {
            consumable,
            writer,
            safeInsert,
            updateConversionResult
          } = conversionApi;
          const link = data.viewItem;
          const linkChild = link.getChild(0);

          // Links that do not wrap an image will be ignored by this plugin
          if (!linkChild?.is( 'element', 'img')) return;

          // Tests if the view elements can be consumed
          if (!consumable.test(link, wrapper)) return;
          if (!consumable.test(linkChild, wrapper)) return;

          // Create model element
          const modelElement = writer.createElement('myimage');

          // Insert element on a current cursor location
          if (!safeInsert(modelElement, data.modelCursor)) return;

          // Consume the element
          consumable.consume(link, wrapper);
          consumable.consume(linkChild, wrapper);
          consumable.add(link, { attributes: ['href', 'target' ] });
          consumable.add(linkChild, { attributes: ['alt', 'src'] });

          writer.setAttributes(linkChild.getAttributes(), modelElement);

          updateConversionResult( modelElement, data );
        });
      });
    }

    editor.conversion.for('downcast').elementToStructure({
        model: 'myimage',
        view: (modelElement, { writer } ) => {
          const linkAttributes = {
            href: modelElement.getAttribute('href'),
            target: modelElement.getAttribute('target')
          }
          const imageAttributes = {
            alt: modelElement.getAttribute('alt'),
            class: modelElement.getAttribute('class'),
            src: modelElement.getAttribute('src'),
            style: modelElement.getAttribute('style')
          }
          if (modelElement.hasAttribute('href')) {
            // a wrapping img
            return writer.createContainerElement('a', linkAttributes, [writer.createAttributeElement('img', imageAttributes)]);
          } else {
            // img only
            return writer.createAttributeElement('img', imageAttributes);
          }
        }
    });
```

**Comment by Witoso** [comment]
Hi @cordesmj, sorry for the late reply, did you manage to find a solution?

**Comment by cordesmj** [comment]
> Hi [@cordesmj](https://github.com/cordesmj), sorry for the late reply, did you manage to find a solution?

Hi @Witoso  - thanks for replying. The only thing I managed to get sort of working was what I posted above in my comment after the original post.
Unfortunately, I found that my `elementToStructure` downcast conversion didn't work as expected. When I removed the `href` attribute from the model, I found that I still had a `<a>` tag (with no attributes) wrapping my `<img>` in the view.
I really wanted to do this with a single model, but I didn't want to have a bunch of useless `<a>` tags cluttering up my content, so I ended up going back to my original approach, having two models, `myImage` and `myLink`. 😖
